### PR TITLE
feature: resolve region from profile and env

### PR DESCRIPTION
**Pull request by stobrien89**
Description of changes:

Adds the ability to configure region via env var and shared config file. This would mean the region is no longer required in the client configuration. This branches from https://github.com/aws/aws-sdk-php/pull/2693, so the changes in question here are _apply_region, _resolve_region some updates to _missing_region. This is an initial draft/proof-of-concept.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/ClientResolver.php [hunk 1]
```diff
@@ -533,10 +534,10 @@ public static function _apply_disable_request_compression($value, array &$args)
             $value = $value();
         }
         if (!is_bool($value)) {
-           throw new IAE(
-              "Invalid configuration value provided for 'disable_request_compression'."
-              . " value must be a bool."
-           );
+            throw new IAE(
```
Comment (SamRemis) [inline src/ClientResolver.php:537]: Hard to see this on GH, but looks like this spacing should be undone

File: src/ClientResolver.php [hunk 2]
```diff
@@ -92,8 +92,9 @@ class ClientResolver
         'region' => [
             'type'     => 'value',
```
Comment (SamRemis) [inline src/ClientResolver.php:93]: Please add a changelog file

File: src/ClientResolver.php [hunk 3]
```diff
@@ -1115,15 +1116,33 @@ public static function _default_signing_region(array &$args)
...
+    }
+
+    public static function _default_region(&$args)
+    {
+        return ConfigurationResolver::resolve('region', '', 'string');
+    }
+
     public static function _missing_region(array $args)
     {
         $service = isset($args['service']) ? $args['service'] : '';
```
Comment (SamRemis) [inline src/ClientResolver.php:1134]: When will service ever be blank?  Is there a good use case for resolving a client without a service using this file?

File: src/ClientResolver.php (hunk 1 above)
Comment (stobrien89) [reply src/ClientResolver.php:537]: This is actually right.  Just verified locally.

**Review by SamRemis — APPROVED** [review]

**Pull request merged**